tokio-util/src/sync/cancellation_token.rs

```diff
@@ -319,560 +190,323 @@ impl<'a> Future for WaitForCancellationFuture<'a> {
+        let num_handles = {
+            let mut locked_node = node.inner.lock().unwrap();
+            locked_node.num_handles -= 1;
+            locked_node.num_handles
         };
 
-        // This task cancelled the token
-
-        // Take the task list out of the Token
-        // We do not want to cancel child token inside this lock. If one of the
-        // child tasks would have additional child tokens, we would recursively
-        // take locks.
-
-        // Doing this action has an impact if the child token is dropped concurrently:
-        // It will try to deregister itself from the parent task, but can not find
-        // itself in the task list anymore. Therefore it needs to assume the parent
-        // has extracted the list and will process it. It may not modify the list.
-        // This is OK from a memory safety perspective, since the parent still
-        // retains a reference to the child task until it finished iterating over
-        // it.
-
-        let mut first_child = {
-            let mut guard = self.synchronized.lock().unwrap();
-            // Save the cancellation also inside the Mutex
-            // This allows child tokens which want to detach themselves to detect
-            // that this is no longer required since the parent cleared the list.
-            guard.is_cancelled = true;
-
-            // Wakeup all waiters
-            // This happens inside the lock to make cancellation reliable
-            // If we would access waiters outside of the lock, the pointers
-            // may no longer be valid.
-            // Typically this shouldn't be an issue, since waking a task should
-            // only move it from the blocked into the ready state and not have
-            // further side effects.
-
-            // Use a reverse iterator, so that the oldest waiter gets
-            // scheduled first
-            guard.waiters.reverse_drain(|waiter| {
-                // We are not allowed to move the `Waker` out of the list node.
-                // The `Future` relies on the fact that the old `Waker` stays there
-                // as long as the `Future` has not completed in order to perform
-                // the `will_wake()` check.
-                // Therefore `wake_by_ref` is used instead of `wake()`
-                if let Some(handle) = &mut waiter.task {
-                    handle.wake_by_ref();
+        if num_handles == 0 {
+            with_locked_node_and_parent(node, |mut node, parent| {
+                // Move all children to parent
+                match parent {
+                    Some(mut parent) => {
+                        move_children_to_parent(&mut node, &mut parent);
+
+                        // Remove the node from the parent
+                        remove_child(&mut parent, node);
+                    }
+                    None => {
+                        disconnect_children(&mut node);
+                    }
```

Comment:
If this happens in parallel with a call to cancel on the parent node, after `node` is removed from its parent, but before `cancel` is called on `node`, then the cancel is lost for the children.

Possible solution: Don't destroy the tree when cancelling.

Reply:
Just for understanding: 
- You call "Cancel" on the parent.
- The parent disconnects the node
- On other thread: the last handle gets dropped. Because the node no longer has a parent, "disconnect_children" gets called
- The "cancel" function finds a node with no children and decides that's it

If with "don't destroy the tree when cancelling" you mean, "First cancel everything, then destroy the tree", I agree.
We just gotta be careful about what happens when we get multiple cancels on different nodes simultaneously.

Related question: Do we actually have to destroy the tree when cancelling? Do we gain anything from it? Originally I thought it might reduce the number of nodes, but now that we implemented that we move children to parents when no handles exist any more, I don't think disconnecting the tree gives us anything any more.

We further have to be careful though, that when we iterate through the tree we don't run into the same problem.

The actual problem is that there is a gap in cancel() between locking the children for disconnect and locking the children for propagating the cancellation. I'll attempt to fix that.

Reply:
I realized while iterating over the tree for cancellation+dropping, we can only iterate over it depth first, because we have to keep the lock of the parent alive during the entire time of iterating through the children, and we cannot lock two children simultaneously. That means we have to descend one child at a time.

Reply:
Fixed, potentially